cuda/base/kernel_launch.cuh

```diff
@@ -0,0 +1,243 @@
+__global__ __launch_bounds__(default_block_size) void generic_kernel_1d(
+    size_type size, KernelFunction fn, KernelArgs... args)
+{
+    auto tidx = thread::get_thread_id_flat();
+    if (tidx >= size) {
+        return;
+    }
+    fn(tidx, args...);
+}
+
+
+template <typename KernelFunction, typename... KernelArgs>
+__global__ __launch_bounds__(default_block_size) void generic_kernel_2d(
+    size_type rows, size_type cols, KernelFunction fn, KernelArgs... args)
+{
+    auto tidx = thread::get_thread_id_flat();
+    auto col = tidx % cols;
+    auto row = tidx / cols;
+    if (row >= rows) {
+        return;
+    }
+    fn(row, col, device_unpack_2d_impl<KernelArgs>::unpack(args, cols)...);
+}
+
+
+}  // namespace cuda
+}  // namespace kernels
+
+
+template <typename ValueType>
+kernels::cuda::compact_dense_wrapper<kernels::cuda::cuda_type<ValueType>>
+compact(matrix::Dense<ValueType> *mtx)
+{
+    GKO_ASSERT(mtx->get_stride() == mtx->get_size()[1]);
+    return {kernels::cuda::as_cuda_type(mtx->get_values())};
+}
+
+
+template <typename ValueType>
+kernels::cuda::compact_dense_wrapper<const kernels::cuda::cuda_type<ValueType>>
+compact(const matrix::Dense<ValueType> *mtx)
+{
+    GKO_ASSERT(mtx->get_stride() == mtx->get_size()[1]);
+    return {kernels::cuda::as_cuda_type(mtx->get_const_values())};
+}
+
+
+template <typename ValueType>
+kernels::cuda::cuda_type<ValueType> *vector(matrix::Dense<ValueType> *mtx)
+{
+    GKO_ASSERT(mtx->get_size()[0] == 1 ||
+               (mtx->get_size()[1] == 1 && mtx->get_stride() == 1));
+    return kernels::cuda::as_cuda_type(mtx->get_values());
+}
+
+
+template <typename ValueType>
+const kernels::cuda::cuda_type<ValueType> *vector(
+    const matrix::Dense<ValueType> *mtx)
+{
+    GKO_ASSERT(mtx->get_size()[0] == 1 ||
+               (mtx->get_size()[1] == 1 && mtx->get_stride() == 1));
+    return kernels::cuda::as_cuda_type(mtx->get_const_values());
+}
+
+
+template <typename KernelFunction, typename... KernelArgs>
+void run_kernel(std::shared_ptr<const CudaExecutor> exec, KernelFunction fn,
+                size_type size, KernelArgs &&... args)
+{
+    cuda::device_guard guard{exec->get_device_id()};
+    constexpr auto block_size = kernels::cuda::default_block_size;
+    auto num_blocks = ceildiv(size, block_size);
+    kernels::cuda::generic_kernel_1d<<<num_blocks, block_size>>>(
+        size, fn, kernels::cuda::map_to_device(args)...);
+}
+
+template <typename KernelFunction, typename... KernelArgs>
+void run_kernel(std::shared_ptr<const CudaExecutor> exec, KernelFunction fn,
+                dim<2> size, KernelArgs &&... args)
+{
+    cuda::device_guard guard{exec->get_device_id()};
+    constexpr auto block_size = kernels::cuda::default_block_size;
+    auto num_blocks = ceildiv(size[0] * size[1], block_size);
+    kernels::cuda::generic_kernel_2d<<<num_blocks, block_size>>>(
+        size[0], size[1], fn, kernels::cuda::map_to_device(args)...);
```

Comment:
Why are breaking `dim<2>` open instead of giving the kernel a single `dim<2>` argument?
If you would give the auther the size into the `KernelFunction`, this would also only occupy a single argument instead of two.

Reply:
I tried to avoid using more complex types on the device. Since this is the backend, where do you see the difference in using a single argument instead of two?

Reply:
Better self-documentation and extensibility (for 3D, 4D, ...).
Also, with a `dim`, it is impossible to accidentally swap rows and cols (otherwise, you could swap them in the kernel call or in the variable naming of the kernel).
But this is obviously a very minor thing and I would be fine to keep it like it is now, but I would prefer `dim<2>`.

Reply:
Since we have individual functions for all of these types of kernels, I would prefer leaving it as it is.